### ISSUE TITLE
Size mismatch in Semantic Segmentation in Torch

### ISSUE THREAD

**Issue by kvrd18** [NVIDIA/DIGITS#1357]
Hi,

I have a dataset of 9000+ `512x512` images with associated labels. I'm defining a simple toy model in Torch for training a semantic segmentation network like this:

```lua
return function(p)
    model = nn.Sequential()
    model:add(nn.SpatialConvolution(1,2,3,3,1,1,1,1))
    return {
        model = model,
        loss = nn.SpatialClassNLLCriterion()
    }
end
```

The dataset is prepared as follows:

        DB backend: lmdb
        Create train_db DB
                Entry Count: 8523
                Feature shape (1, 512, 512)
                Label shape (1, 512, 512)
        Create val_db DB
                Entry Count: 947
                Feature shape (1, 512, 512)
                Label shape (1, 512, 512)

I get the following error:

	ERROR: /home/predible/torch/install/share/lua/5.1/nn/THNN.lua:109: bad argument #1 to 'v' (only batches of spatial targets supported (3D tensors) at /home/predible/Kiran/XrayFracture/LIB/cunn/lib/THCUNN/SpatialClassNLLCriterion.cu:96)

Is there a way to reshape the input and target tensors before feeding into the `model:forward()` and `model:backward()` in DIGITS?

**Comment by kvrd18** [comment]
The following snippet works perfectly for Semantic Segmentation in Torch:

```lua
require 'nn';

x = torch.rand(1,1,512,512)
y = torch.ones(1,512,512)

model = nn.Sequential()
model:add(nn.SpatialConvolution(1,2,3,3,1,1,1,1))
criterion = nn.SpatialClassNLLCriterion()

out = model:forward(x)
loss = criterion:forward(x,y)
gradIn = criterion:backward(x,y)
x_in = model:backward(x,gradIn)
```

**Comment by gheinrich** [comment]
Hello, why don't you reshape your tensors within the network?

**Comment by kvrd18** [comment]
`nn.SpatialClassNLLCriterion()` expects a 4D tensor as input and 3D tensor as target for its `:forward() and :backward(). I'm not sure if DIGITS allows that.

**Comment by gheinrich** [comment]
You can reshape the target using a [`labelHook`](https://github.com/NVIDIA/DIGITS/blob/digits-5.0/docs/GettingStartedTorch.md#internal-parameters). I have a pending change to extend this to all data (input and target).

In the meantime you can reshape the input with a `nn.Reshape` layer. Have a look at https://github.com/torch/nn/blob/master/doc/simple.md#nn.Reshape

**Comment by kvrd18** [comment]
Thank you. [`labelHook`](https://github.com/NVIDIA/DIGITS/blob/digits-5.0/docs/GettingStartedTorch.md#internal-parameters) works. I shall close the issue here.
